### PR TITLE
[TG-7815] Fix bug in get_recursively_instantiated_type

### PR DESCRIPTION
**Pull request by NathanJPhillips**
This reduces the amount of code and in so doing fixes the implementation of specialization of generics.

Previously the following invariant was being triggered in non-symmetric mutual generic specialization:
```
Invariant check failed
File: .../lib/cbmc/jbmc/src/java_bytecode/select_pointer_type.cpp:187 function: get_recursively_instantiated_type
Condition: depth < replacements.size()
Reason: cannot access elements outside stack
```

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- N/A The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included in a downstream PR.
- N/A My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 1]
```diff
@@ -126,86 +126,24 @@ pointer_typet select_pointer_typet::specialize_generics(
...
-    return {};
-  }
-
-  const size_t index = (replacements.size() - 1) - depth;
-  const auto &type = replacements[index];
-
-  if(!is_java_generic_parameter(type))
-  {
-    return type;
+    auto types_it = generic_parameter_specialization_map.find(parameter_name);
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp]: Commit message: note what the bug was and how this fixes it

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 2]
```diff
@@ -126,86 +126,24 @@ pointer_typet select_pointer_typet::specialize_generics(
...
+  generic_parameter_specialization_mapt
+    generic_parameter_specialization_map) const
 {
-  generic_parameter_recursion_trackingt visited;
-  const size_t max_depth =
-    generic_parameter_specialization_map.find(parameter_name)->second.size();
-
-  irep_idt current_parameter = parameter_name;
-  for(size_t depth = 0; depth < max_depth; depth++)
+  while(true)
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp]: Looks good, but since taking a map by value like this is quite unusual please add comments noting the strategy you're using here and why

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 3]
```diff
@@ -82,26 +81,27 @@ pointer_typet select_pointer_typet::specialize_generics(
...
-          to_pointer_type(array_element_type),
-          generic_parameter_specialization_map,
-          visited_nodes);
-
-        pointer_typet replacement_array_type = java_array_type('a');
-        replacement_array_type.subtype().set(ID_element_type, new_array_type);
-        return replacement_array_type;
-      }
-    }
+    auto array_element_type =
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:80]: `const auto *` for clarity?

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 4]
```diff
@@ -82,26 +81,27 @@ pointer_typet select_pointer_typet::specialize_generics(
...
+    const pointer_typet &new_array_type = specialize_generics(
+      *array_element_type,
+      generic_parameter_specialization_map,
+      visited_nodes);
+
+    pointer_typet replacement_array_type = java_array_type('a');
+    replacement_array_type.subtype().set(ID_element_type, new_array_type);
+    return replacement_array_type;
   }
+
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:93]: Don't re-space in the same commit as a substantial change

**Review by smowton — COMMENTED** [review]
Also please add at least one regression test exposing the bug (or better yet, KNOWNBUG -> CORE an existing test), and there seem to be real unit test failures in CI.

File: jbmc/src/java_bytecode/select_pointer_type.cpp (hunk 2 above)
Comment (allredj) [reply jbmc/src/java_bytecode/select_pointer_type.cpp]: The map itself is not taken by value. I don't think we need to document why passing a struct by value is OK.

File: jbmc/src/java_bytecode/java_types.h [hunk 5]
```diff
@@ -735,13 +735,23 @@ class java_generic_parametert : public reference_typet
...
+}
+
 /// Checks whether the type is a java generic parameter/variable, e.g., `T` in
 /// `List<T>`.
 /// \param type: the type to check
 /// \return true if type is a generic Java parameter type
 inline bool is_java_generic_parameter(const typet &type)
 {
-  return is_reference(type) && is_java_generic_parameter_tag(type.subtype());
+  return can_cast_type<java_generic_parametert>(type);
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_types.h:754]: Should probably be in a separate commit. It does not correspond to the commit description. There are several things being changed here, and it makes it difficult to check that this commit does not introduce a behavioural change.

File: jbmc/src/java_bytecode/select_pointer_type.h [hunk 6]
```diff
@@ -23,6 +24,32 @@ typedef std::unordered_map<irep_idt, std::vector<reference_typet>>
...
 
+struct print_generic_parameter_specialization_map
+{
+  const generic_parameter_specialization_mapt &map;
+  const namespacet &ns;
+};
+
+template <typename ostream>
+ostream &
+operator<<(ostream &stm, const print_generic_parameter_specialization_map &map)
```
Comment (majakusber) [inline jbmc/src/java_bytecode/select_pointer_type.h:34]: :ok_hand: This is useful

**Review by allredj — COMMENTED** [review]
Should be fine once the tests are in.

File: jbmc/src/java_bytecode/select_pointer_type.cpp (hunk 2 above)
Comment (majakusber) [reply jbmc/src/java_bytecode/select_pointer_type.cpp]: I'm not sure what you mean by "not taken by value". We do make a copy here, so we can pop from the stacks of particular parameters in the map just for the sake of finding the parameter specialization, but at the call site of this function we need to keep the original full information. I agree that adding a note that points this out is useful (not arguing why it's ok, just to make the reader notice), I missed it myself at first.

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 7]
```diff
@@ -25,61 +25,60 @@ pointer_typet select_pointer_typet::convert_pointer_type(
...
+/// - T -> (Integer, U)
+/// - U -> (String, T)
+///
+/// Note that the top of the stacks for T and U create a recursion T -> U,
+/// U-> T. We want to break it and specialize `T my.gen.t` to `String my.gen.t`.
+/// \param pointer_type: The pointer type to specialize
+/// \param generic_parameter_specialization_map: Map of type names to
+///   specialization stack
+/// \return The first instantiated type for the generic type or nothing if no
+///   such instantiation exists.
```
Comment (majakusber) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:43]: This should be updated slightly - it sounds l like it's only handling the recursion case (because it's just moved from the function that was breaking recursion). Also, move the documentation to the header.

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 8]
```diff
@@ -25,61 +25,60 @@ pointer_typet select_pointer_typet::convert_pointer_type(
...
   if(parameter != nullptr)
   {
-    const irep_idt &parameter_name = parameter->get_name();
+    irep_idt parameter_name = parameter->get_name();
 
-    // avoid infinite recursion by looking at each generic argument from
-    // previous assignments
-    if(visited_nodes.find(parameter_name) != visited_nodes.end())
+    generic_parameter_specialization_mapt spec_map_copy =
+      generic_parameter_specialization_map;
```
Comment (majakusber) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:55]: :bulb: You're making a copy of the map over and over again, changing a single element in it in each call. I don't have any good idea how to make this less costly though.. not so nice ideas being keeping two functions, make the code uglier at the call site, or add extra arguments to the function (depth of each stack to look at). Can you think of a better way?

**Review by majakusber — COMMENTED** [review]
Needs some cleanup but I love it :slightly_smiling_face:

**Comment by NathanJPhillips** [comment]
This solution was inadequate. Replacing with a new solution.
